### PR TITLE
Analyzer: Handle more code tokens

### PR DESCRIPTION
**Pull request by brbrr**
This PR adds support for some missing cases, which was preventing the script from running against some of the plugins

#### Jetpack product discussion

pd5faL-1O-p2
#### Does this pull request change what data or activity we track or use?

n/a

#### Testing instructions:

* It is a bit hard to test it
* Download Facebook to Woo plugin: https://wordpress.org/plugins/facebook-for-woocommerce/
* Extract it somewhere.
* run below script from within `packages/analyzer/scripts` folder:
```
require dirname( __DIR__ ) . '/vendor/autoload.php';
$path_to_plugin = "/abs/path/to/plugin";
$invocations = new Automattic\Jetpack\Analyzer\Invocations();
$invocations->scan( $path_to_plugin);
```
* Make sure there is no warning or notices generated

### CODE REVIEW

File: projects/packages/analyzer/src/class-utils.php [hunk 1]
```diff
@@ -31,24 +31,37 @@ static function get_param_default_as_string( $default, $current_class ) {
...
-			$class_name =
-						'$'
-						. self::maybe_stringify( $node->var->name )
-						. '[' . self::maybe_stringify( $node->dim->value )
-						. ']';
+			$dim_val = '';
+			if ( $node->dim instanceof Node\Expr\Variable ) {
+				$dim_val = '[$' . self::maybe_stringify( $node->dim->name ) . ']';
+			} elseif ( $node->dim instanceof Node\Expr\PropertyFetch ) {
+				$dim_val = '[$' . self::maybe_stringify( $node->dim->var->name ) . ']';
```
Comment (anomiex) [inline projects/packages/analyzer/src/class-utils.php:58]: Is this right? It looks like it should be something more like what's going on in lines 49–51.

For that matter, maybe there should just be a recursive call here (and in some of the other cases)?

File: projects/packages/analyzer/src/class-utils.php [hunk 2]
```diff
@@ -31,24 +31,37 @@ static function get_param_default_as_string( $default, $current_class ) {
...
+			} elseif ( $node->dim instanceof Node\Expr\ArrayDimFetch ) {
+				$dim_val = '[$' . self::maybe_stringify( $node->dim->var->name ) . '["' . self::maybe_stringify( $node->dim->dim->value ) . '"]]';
+			} else {
+				$dim_val = '[' . self::maybe_stringify( $node->dim->value ) . ']';
```
Comment (anomiex) [inline projects/packages/analyzer/src/class-utils.php:62]: Should the output string here be quoted, like you're doing in the `getArrayToken()` being added in the other file?

File: projects/packages/analyzer/src/Invocations/class-visitor.php [hunk 3]
```diff
@@ -56,20 +52,56 @@ public function enterNode( Node $node ) {
...
 	public function leaveNode( Node $node ) {
 	}
+
+	/**
+	 * Recursively generates a (multidimentional) array invocation token.
+	 *
+	 * @param Node $node node object.
+	 * @return string
+	 */
+	private function getArrayToken( Node $node ) {
```
Comment (anomiex) [inline projects/packages/analyzer/src/Invocations/class-visitor.php:88]: What's the difference between this and `Utils::node_to_class_name()`? It seems like you're trying to do basically the same thing, just with handling of different edge cases (that better handles complex stuff in the 'dimension' part, while this handles complex stuff in the 'var' part).

For that matter, "node_to_class_name" doesn't seem to really describe what that method is doing. There are some bits that seem like they're handling a "class name", but other bits that seem to be generically stringifying a variable reference expr.

File: projects/packages/analyzer/src/class-utils.php [hunk 4]
```diff
@@ -31,24 +31,37 @@ static function get_param_default_as_string( $default, $current_class ) {
...
+	 * Gets string representation of the Node's class
+	 *
+	 * @param Node $node Node.
+	 * @param any  $class_for_self Class reference.
+	 * @return string
+	 */
+	public static function node_to_class_name( $node, $class_for_self = null ) {
 		if ( $node instanceof Node\Expr\Variable
 			|| $node instanceof Node\Stmt\Class_ ) {
 			$class_name = $node->name;
```
Comment (anomiex) [inline projects/packages/analyzer/src/class-utils.php:44]: When it's `Node\Expr\Variable`, should the result actually be `'$' . self::maybe_stringify( $node->name )` like I see is being done elsewhere?

**Review by anomiex — COMMENTED** [review]
A few things inline. Stop me if I'm looking too deep into the rabbit hole.

File: projects/packages/analyzer/src/class-utils.php (hunk 2 above)
Comment (brbrr) [reply projects/packages/analyzer/src/class-utils.php:62]: Good call. But I'm not sure tbh :D Quotes were omitted in the initial implementation (line 50 before my changes). I need to dig deeper into the `Differences` APIs to better understand how things work. But I'd say the quoting is out of scope of this PR

File: projects/packages/analyzer/src/class-utils.php (hunk 1 above)
Comment (brbrr) [reply projects/packages/analyzer/src/class-utils.php]: Indeed, it wasn't properly constructing the call. I updated similarly to 49-51

Comment (brbrr) [reply projects/packages/analyzer/src/class-utils.php]: As for refactoring it, I guess it makes sense. But I'd like to address it in further PRs

**Review by anomiex — APPROVED** [review]
I'm ok with this being merged as-is, with potential future fixes in followups. Disabling auto-merge in case you did want to do anything with the other two comments in this PR.

**Pull request merged**